feat(swap): partial approve

shoom3301 opened the pull request:
# Summary

Many users are asking for partial approve in swap because it might be dengerous to give an "infinite" approval amount.
Now we have an option in Swap settings which enables the partial approve feature.
When the option is on, then we will make an approval only for selling amount.

Important! When order is a buy order, then approval amount should be sell amount including all costs (fees, slippage, etc.).

Important! The option works only in Swap. I know, we have approvals in limit orders, tokens page, etc. But for now the feature is availabe only in Swap because it's not clear how it will affect other flows.

<img width="552" alt="image" src="https://github.com/user-attachments/assets/817cc7a1-79b7-4f30-8b12-cefa6544734a" />


# To Test

1. Enable the option in Swap settings
2. Reset allowance for the sell token (https://revoke.cash/)
3. Post an order
- [ ] for regular approve there should be an approval transaction with value == sell amount
- [ ] for permittable token there should be a permit with value == sell amount
- [ ] after order fill the sell token allowance should be still zero (or a small amount for buy orders)

elena-zh commented:
The build has failed :(

shoom3301 commented:
@elena-zh I pushed some fixes:
 - use permit cache only when sell amount exact the same with cached
 - fixed partial approvals in Safe (approve+swap, wrap+approve+swap)

elena-zh commented:
Hey @shoom3301 , great job!

Some questions/issues I have: 
1. (to discuss): for a buy order, we include fees and a slippage amount into a total approval amount. And it is sent to the wallet to approve. But sometimes it may take a bit longer than needed to approve a token, and during this time a price/dynamic slippage may change. It will lead to 'not enough' approval case again on the UI, however, nothing has been changed on the form (input amounts). This may bring a user to confusion ('Why I should approve once again if I've already approved it?'). Might it be possible to increase an approval amount by a certain percent to avoid this case?
![image](https://github.com/user-attachments/assets/6660bf5e-b3ab-4ed4-84e3-2d2fe918d5ae)

2. Hooks: partial approvals are working there. I'm fine with this. So i'd like to double check it with you: should we still disable them on the UI? If so, it would be great to hide this setting from the menu there.
![hooks hide](https://github.com/user-attachments/assets/37ee8327-a33b-46d8-9123-498b09efd157)
![hooks usdt](https://github.com/user-attachments/assets/47be098e-0513-4bf5-a18d-0bf8687726e4)

3. Classic ETH-flow: setting does not affect it. When I run an approval from this flow, max amount to approve is sent to the wallet. Same happens when I try placing ETH order being connected to Safe WC
![eth-flow](https://github.com/user-attachments/assets/563f27ec-6145-4fa2-9a3d-7016e43781a2)
4. (not sure about it, might be some BE issues were involved) Some of my orders with a partial approval hook were not filled :( 
5. Interaction with the Limit orders flow. is this OK when:
 - I sign a full permit on Limit orders page, the order is pending
 - Then go to the swap where 'partial approval' option is switched ON
 - specify an order with the same token --> the app asks me to 'resign' the permit (is this OK?)
  - swap order is executed
  - I go to the limit orders tab, place one more order --> the app asks to sign unlimited approval once again. 
  My concern here: we make a user pay multiple times for approval of the same token :( 
  6. (related to 5) permit hook is displayed on the Confirm modal when:
  - I sign a full permit on Limit orders page, the order is pending
 - Then go to the swap where 'partial approval' option is switched ON
 - specify an order with the same token , open the confirm modal --> a hook is displayed there, but the app asks me to resign permit once again
 7.  I can see 'failed simulation' log on this hook (it must be related to this bug #5185 , but it is visible especially for failed simulations)
![simulation](https://github.com/user-attachments/assets/b3fa6c15-b2df-4f5b-8b65-3a92682c6b9f)

Thanks!

shoom3301 commented:
@elena-zh thank you!

1. I think we shold keep it as it is. Yes, price might move, but user can adjust the sell amount if they need.
2. Fixed, partial approves should work only in Swap.
3. Fixed
4. I don't see any problems on frontend side, need to debug with backenders
5. this is a goal of this feature, to make users able to trade with a specific allowance amount. So, this is ok
6. Fixed
7. I removed partial approvals from Hooks store

elena-zh reviewed (approved):
Hey @shoom3301 , great job!
I'm approving the PR.
Please don't forget to discuss with the team the 1st case reported in https://github.com/cowprotocol/cowswap/pull/5256#issuecomment-2561950483

And just FYI: I'm still facing issues with orders execution ([1](https://dev.explorer.cow.fi/gc/orders/0x7b3f6440399c4c7524fb7124544eed284b65414eb4a00b3b977307c1557284e59fa3c00a92ec5f96b1ad2527ab41b3932efeda58676d30ff?tab=overview), [2](https://dev.explorer.cow.fi/gc/orders/0x93cbdd9acddcaf51466cbfe2b17764c3835684cb7fda74a3749ae893cf32b01e9fa3c00a92ec5f96b1ad2527ab41b3932efeda58676d3170?tab=overview), [3](https://dev.explorer.cow.fi/gc/orders/0x3682dcb7143f5fd10d37e4d7e87e6e74119633cd1d212e517c7b31595698e8ff9fa3c00a92ec5f96b1ad2527ab41b3932efeda58676d3192?tab=overview)) on Gnosis chain (only). However, it might be related to a poor competition that we have on staging. I will retest this on Prod
